subsys/bluetooth/services/cts.c

```diff
@@ -0,0 +1,177 @@
+	/* get unit timestamp from datetime */
+	ts_val.tv_sec = timeutil_timegm(&date_time);
+	if (ts_val.tv_sec == ((time_t)-1)) {
+		return -EOVERFLOW;
+	}
+	LOG_INF("CTS Write Unix: %lld", ts_val.tv_sec);
+
+	/* convert to millis with fraction 256 */
+	ts_val.tv_nsec = ct->fractions256;
+	ts_val.tv_nsec *= NSEC_PER_SEC;
+	ts_val.tv_nsec /= 255;
+
+	return bt_cts_settime(&ts_val, BT_CTS_UPDATE_REASON_MANUAL);
+}
+
+static inline int prepare_current_time_msg(struct bt_cts_time_format *ct,
+					   enum bt_cts_update_reason update_reason)
+{
+	int err;
+
+	err = clock_gettime(CLOCK_REALTIME, &ts_val);
+	if (err) {
+		return err;
+	}
+
+	LOG_INF("CTS Read Unix: %lld", ts_val.tv_sec);
+	/* generate date time from unix timestamp */
+	if (gmtime_r(&ts_val.tv_sec, &date_time) == NULL) {
+		return -EOVERFLOW;
+	}
+	date_time.tm_year += TIME_UTILS_BASE_YEAR;
+
+	LOG_INF("CTS Read Time: %d/%d/%d %d:%d:%d", date_time.tm_year, date_time.tm_mon,
+		date_time.tm_mday, date_time.tm_hour, date_time.tm_min, date_time.tm_sec);
+
+	/* 'Exact Time 256' contains 'Day Date Time' which contains
+	 * 'Date Time' - characteristic contains fields for:
+	 * year, month, day, hours, minutes and seconds.
+	 */
+	ct->year = sys_cpu_to_le16(date_time.tm_year);
+	ct->mon = date_time.tm_mon + 1; /* months starting from 1 */
+	ct->mday = date_time.tm_mday;   /* Day of month */
+	ct->hours = date_time.tm_hour;  /* hours */
+	ct->min = date_time.tm_min;     /* minutes */
+	ct->sec = date_time.tm_sec;     /* seconds */
+	/* day of week starting from 1-monday, 7-sunday */
+	ct->wday = date_time.tm_wday;
+	if (ct->wday == 0) {
+		ct->wday = 7; /* sunday is represented as 7 */
+	}
+	/* 'Fractions 256 part of 'Exact Time 256' */
+	ts_val.tv_nsec *= 255;
+	ts_val.tv_nsec /= NSEC_PER_SEC;
+	ct->fractions256 = ts_val.tv_nsec;
+	/* Adjust reason */
+	ct->reason = update_reason;
+	return 0;
+}
+
+static void ct_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	ARG_UNUSED(attr);
+
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+
+	LOG_INF("CTS Notifications %s", notif_enabled ? "enabled" : "disabled");
+}
+
+static ssize_t read_ct(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+		       uint16_t len, uint16_t offset)
+{
+	int err = prepare_current_time_msg(&ct_time, BT_CTS_UPDATE_REASON_UNKNOWN);
+
+	if (!err) {
+		return bt_gatt_attr_read(conn, attr, buf, len, offset, &ct_time, sizeof(ct_time));
+	} else {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+}
+
+static ssize_t write_ct(struct bt_conn *conn, const struct bt_gatt_attr *attr, const void *buf,
+			uint16_t len, uint16_t offset, uint8_t flags)
+{
+	int err;
+
+	if ((offset != 0) || (offset + len != sizeof(ct_time))) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	memcpy(&ct_time, buf, sizeof(ct_time));
+	err = calculate_ms_ref_from_current_time_msg(&ct_time);
+
+	if (!err) {
+		return len;
+	} else {
+		return BT_GATT_ERR(0x80);
```

Comment:
what's the 0x80?

Reply:
as per CTS_SPEC_V1.1.0.pdf, section 1.6, server need to send 0x80 error code in case one or more fields are ignored